From f24230e9e83537ee0f85138d5840ef4886254711 Mon Sep 17 00:00:00 2001
From: "Maneendra,Pokuri Naga" <pnmaneendra@maxlinear.com>
Date: Mon, 14 Oct 2024 15:49:19 +0530
Subject: [PATCH] WLANRTSYS-85028 To avoid MLD struct deinit before even
 setting up the MLD VAP Issue: 1.While creating MLD, first both VAPs will
 update the global p_ap_mld structers one after the other. 2.But current
 problem is after updating the structures, even before setting up the MLD VAPs
 based on the global p_ap_mld structure   Stop hostapd from pwhm is causing
 MLD global struct deinitialization which is clearing the previously filled
 VAP details.

Fix:
1.Added Check if 2 Links available in p_ap_mld structs without ML VAP creation and interface is going down avoid calling MLD vap structure deinit
2.and also one more problem is when a MLD is getting deinitialized, we are not resetting the mld_id parameter of the hapd conf to default value.

Fixup to WLANRTSYS-60843 MLO changes: AP MLD configuration.
---
 src/ap/mxl_hostapd.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/src/ap/mxl_hostapd.c b/src/ap/mxl_hostapd.c
index a34a0d89d..02832c373 100644
--- a/src/ap/mxl_hostapd.c
+++ b/src/ap/mxl_hostapd.c
@@ -2698,7 +2698,7 @@ void mxl_hostapd_remove_mld_vap(struct hostapd_iface *iface, unsigned int idx)
 
 int mxl_hostapd_deinit_gmld(struct hostapd_data *hapd)
 {
-	struct p_ap_mld *gmld =  NULL;
+	struct p_ap_mld *gmld = NULL;
 	int mld_id;
 
 	if (hapd == NULL)
@@ -2707,14 +2707,18 @@ int mxl_hostapd_deinit_gmld(struct hostapd_data *hapd)
 	if (hapd->conf->mld_id == INVALID_MLD_ID)
 		return 0;
 
+	mld_id = hapd->conf->mld_id;
+	gmld = &(hapd->iface->interfaces->mxl_interfaces.g_ap_mld[mld_id]);
+	if (gmld->num_links == MAX_NUM_LINKS)
+		return 0;
+
 	if (hapd->conf->mld_id >= MAX_SUPPORTED_MLDS) {
 		wpa_printf(MSG_ERROR, "Invalid MLD id %d. Cannot deinit gmld", hapd->conf->mld_id);
 		return -1;
 	}
 
-	mld_id = hapd->conf->mld_id;
-	gmld = &(hapd->iface->interfaces->mxl_interfaces.g_ap_mld[mld_id]);
 	os_memset(gmld, 0, sizeof(struct p_ap_mld));
+	hapd->conf->mld_id = INVALID_MLD_ID;
 
 	wpa_printf(MSG_INFO, "MLD id %d deinitialized", hapd->conf->mld_id);
 
-- 
2.43.0

